sdk/python/kfp/components/_components.py

```diff
@@ -186,156 +186,50 @@ def _make_name_unique_by_adding_index(name:str, collection, delimiter:str):
     return unique_name
 
 
+#Holds the transformation functions that are called each time TaskSpec instance is created from a component. If there are multiple handlers, the last one is used.
+_created_task_transformation_handler = []
```

Comment:
Can you add some comments about _created_task_transformation_handler? The only place this is populated seems to be the line 195 so it is not clear why it needs to exist as a list.

Reply:
The intent is to set it in the `Pipeline` context. I had that code in the PR, but then I decided to split that into its own PR. The intent will be more obvious there.
**Python context managers that replace some value usually make the value holder a list so that the context is re-entrant** - so that the code does not break if you happen to have nested contexts. https://docs.python.org/3/library/contextlib.html#reentrant-cms

The high-level idea is as follows: Conceptually, what you get when you give arguments to a Component is a Task. But currently we need to get a `ContainerOp` instance. So there needs to be a `TaskSpec` -> `ContainerOp` transformation that's applied automatically when the pipeline is being compiled.

Reply:
What do you think would be the best wording for a comment?

Reply:
In some new cases (the existing cases will continue to work) `load_component` will need to return  `TaskSpec` objects instead of `ContainerOp` objects. So there needs to be a way to enable/disable the `TaskSpec` conversion to `ContainerOp`. The `_created_task_transformation_handler` holds the transformation procedure that can be changed or disabled.

Reply:
Since some of the code is for graph components, would you hold it off this when we decide graph component priority? Or, is it possible to decouple graph component from container component?

Reply:
This is blocking the whole components and intermediate YAML efforts.
This blocks the backend component API (@vicaire, @IronPan).
This blocks passing the input/output metadata for the UI. (@gaoning777 @rileyjbauer @yebrahim @IronPan)
This blocks the artifacts support.

Reply:
This code change is a foundation that is needed for multiple puzzle pieces that I'm asked to deliver. It's needed for the following efforts:
* Debug a single component locally (container components for now, graph components - later)
* Submit a run for a single component which is needed for component tests (container components for now, graph components - later)
* Intermediate YAML API in the backend.
* Passing the pipeline metadata (description) and input/output metadata (names, descriptions, types) to the UI. After discussing with @vicaire, @IronPan and @gaoning777 it was decided to skip the "pass information through Argo template metadata" short-term approach I suggested and either go with the intermediate YAML only or attach that intermediate yaml to the end of Argo YAML.
* Implementing static type checking - All 6 cases (submission from UI, submission from Python, constant arguments in python pipeline, passing outputs to inputs when composing pipelines in Python, passing outputs to inputs when composing pipelines in graph component YAML)
* Artifact support. Artifacts are involved when passing data between components which happens in a graph and requires the graph support.

The reason this affects that much is that many features depend on having the full information which `TaskSpec` has, but that's lost in transition to `ContainerOp`.

Reply:
Just to clarify: The actual change that will make "component task factory function" output `TaskSpec` instead of `ContainerOp` in some scenarios is coming in the next PR.

This PR is just a refactoring that does **not** change the behavior.

BTW, this was the intent behind the whole `_dsl_bridge.py` file and the `_task_object_factory` variable - bridge between the component structures and dsl structures like `ContainerOp` https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/components/_dsl_bridge.py#L40 . It turned out the initial effort was insufficient since only limited information was passed to that handler.